### DISCUSSION TITLE
Help with running HSS in Renode with PolarFire SoC Icicle Example .resc files.

### DISCUSSION THREAD

**Discussion by sastel** [polarfire-soc/polarfire-soc-documentation#369] [General]
# Summary
Greetings dear readers.

I'm just getting into PolarFire SoC, and I am trying to run a custom build of the HSS in Renode. Earlier this week I was able to run the default `icicle-kit.resc` from renode with no issues. This script runs the HSS (0.99.7), U-Boot, and Linux. I also hacked up the icicle renode examples to launch baremetal NuttX without the HSS, and was greeted by the nsh prompt. So far so good.

The next step is to try to build the HSS to launch Nuttx on Renode. When I tried to build HSS from v2023.6, I ran into multiple issues culminating in MMC related failures. I'm hoping someone help me to understand if I'm making mistakes, or if I'm dealing with an HSS or Renode problem.

# The Setup
## Versions
- Renode version 1.14.0.29506
- HSS v2023.6

## HSS Build Commands:
```
make BOARD=mpfs-icicle-kit-es clean
make BOARD=mpfs-icicle-kit-es defconfig
make BOARD=mpfs-icicle-kit-es V=1 -j8
```
## My Renode Scripts
I have not modified any of the underlying scripts from renode, only `icicle-kit.resc`. I am tracking my changes in a local git repo. Here is the version I have hacked up for my testing.
I don't think the payload matters at this point, because the issues I'm encountering are occurring in the HSS.
`icicle-kit.resc`:
```
:name: Icicle Kit
:description: This is a sample script running HSS and NuttX on Icicle Kit with PolarFire SoC

#$bin?=@https://dl.antmicro.com/projects/renode/icicle--hss.elf-s_3297936-bcb7ef60abc78a878995554160eaac1dea962e95
$bin?=@/home/sstel/workspace/hart-software-services/Default/hss-envm-wrapper.elf
#$bin?=@/home/sstel/workspace/hart-software-services/Default/hss-l2scratch.elf

$emmc?=@https://dl.antmicro.com/projects/renode/icicle--emmc.img-s_26746880-3a6ef871bc8eb6fcfbda344e8c23fb534ef89961
#$emmc?=@/home/sstel/workspace/nuttxspace/nuttx/payload.bin

i @/home/sstel/workspace/renode/polarfire-soc.resc

# these settings guarantee a deterministic, successful boot-up of Linux
emulation SetGlobalSerialExecution true
emulation SetGlobalQuantum "0.0008"

showAnalyzer mmuart1
machine SdCardFromFile $emmc mmc 0x100000000
```
`polarfire-soc.resc`:
```
:name: PolarFire SoC
:description: This is a sample script prepared to create a PolarFire SoC platform

$bin?=@https://dl.antmicro.com/projects/renode/pse-freertos-lwip.elf-s_1413416-f348e11b20bd1a8cb24c39e888cf6430e4ee4035

using sysbus
mach create

machine LoadPlatformDescription @platforms/cpus/polarfire-soc.repl

showAnalyzer mmuart0

macro reset
"""
   sysbus LoadELF $bin
"""
runMacro $reset

## Alternatively, to start GDB and load from SoftConsole:
# machine StartGdbServer 3333 true
## Remember! After the connection is done and the binary is loaded, you have to issue:
# u54_1 PC `e51 PC`
## for every core you need to use as GDB initializes only the e51 core.
## If you don't want the core to run, issue:
# u54_4 IsHalted true
```
# Issues
## 1. DDR Training
DDR Training hangs at 80%. The solution to this was to set `CONFIG_SKIP_DDR=y` in `boards/mpfs-icicle-kit-es/def_config` in the HSS project, then rebuild.
Is it expected behaviour that a renode sim will hang at 80% DDR training?

## 2. DDR Init Assertion
![image](https://github.com/polarfire-soc/polarfire-soc-documentation/assets/33885247/bb3bc471-ba1f-4eff-bdab-bf06b3d1809a)
```
bool HSS_DDRPrintL2CacheWaysConfig(void)
{
    mHSS_DEBUG_PRINTF(LOG_STATUS, "L2 Cache Configuration:\n");

    const unsigned int way_enable = CACHE_CTRL->WAY_ENABLE + 1u;
    const unsigned int way_mask = (1u << way_enable) - 1u;
    const unsigned int cache_ways = __builtin_popcount(CACHE_CTRL->WAY_MASK_E51_DCACHE & way_mask);
    const unsigned int scratch_ways = way_enable - cache_ways;
    const unsigned int lim_ways =  (unsigned int)(16u - way_enable);

    assert(way_enable <= 16u);
    assert(cache_ways <= 16u);

    mHSS_DEBUG_PRINTF_EX("    L2-Scratchpad: % 2u way%c (%u KiB)\n", scratch_ways, scratch_ways==1u ? ' ':'s', scratch_ways * 128u);
    mHSS_DEBUG_PRINTF_EX("         L2-Cache: % 2u way%c (%u KiB)\n", cache_ways, cache_ways==1u ? ' ':'s', cache_ways * 128u);
    mHSS_DEBUG_PRINTF_EX("           L2-LIM: % 2u way%c (%u KiB)\n", lim_ways, lim_ways==1 ? ' ':'s', lim_ways * 128);

    // sanity check on L2 Size and LIM size...
    // to ensure Libero and the HSS linker script match
    extern const uint64_t __l2_start;
    extern const uint64_t _hss_start;
    extern const uint64_t _hss_end;
    const uintptr_t libero_l2_end = (uintptr_t)&__l2_start + (scratch_ways * 128u * 1024u);

    assert(&_hss_start == &__l2_start);
    assert((uintptr_t)&_hss_end <= libero_l2_end);

    return true;
}
```

The values I saw were:
```
&_hss_start == 0xa000000 == &__l2_start
&_hss_end == 0xa021744
libero_l2_end == 0xa020000
```

The comment above the assert in the code implies that this is an issue with the linker scripts. I'm guessing it really means the HSS has been configured with too many features for the available L2 space. I was able to shrink it down by fiddling somewhat ignorantly with the configuration. I turned off the following:

- `CONFIG_SERVICE_BEU`
- `CONFIG_SERVICE_MMC_MODE_SDCARD`
- `CONFIG_USE_PCIE`
- `CONFIG_USE_IHC`
- `CONFIG_COLOR_OUTPUT`
- `CONFIG_USE_LOGO`
- `CONFIG_DEBUG_PERF_CTRS`
- `CONFIG_DEBUG_PERF_CTRS_NUM`

Am I on the right track here, and if so, what would you suggest removing? 

## 3. `Attempting to select eMMC ... Failed`
This is as far as I've gotten so far. This seems to be a problem with the HSS and renode.

`HSS_MMCInit()` is returning 0, and this is because mmc_init_emmc() fails. MSS_MMC_init() is returning `2 == MSS_MMC_NOT_INITIALIZED`. This gets into the depths of `mss_mmc.c`. I'm wondering if the HSS software is making assumptions about the MMC that aren't valid in the renode emulation.

Does anyone have any suggestions for what to do next?

Wow, that's a lot of information. I've tried to describe it well. Hopefully it is concise enough and not too verbose. Thanks for making it to the end of this post!

**Comment by sastel** [comment]
# Update 

## MMC Bus Voltage Config

With the help of gdb this morning I found the problem. `mss_mmc.c::MSS_MMC_init()` is returning `MSS_MMC_NOT_INITIALIZED` because when it calls `set_sdhost_power(SRS10_SET_1_8V_BUS_VOLTAGE)` the call returns `MSS_MMC_INVALID_PARAMETER`.

`SRS16_VOLTAGE_3_3V_SUPPORT` was set in `MMC->SRS16`. So I reconfigured the HSS to use 3V3 for the MMC bus voltage by removing `CONFIG_SERVICE_MMC_BUS_VOLTAGE_1V8`.

## mmccard_oper_config()

The next problem is that `mmccard_oper_config()` is not returning `MSS_MMC_CARD_SELECT_SUCCESS`. Will work on this when I get a chance.

**Reply by sarah18-z** [reply]
Hello, I encountered the same error when attempting to compile the HSS to launch the FreeRTOS application on Renode. Could you please provide me with information on the solution you attempted to adopt for this issue and what exactly caused it? Thanks!

**Comment by sastel** [comment]
I can build v2023.09 HSS with the default icicle config, and run it on a physical icicle board without any issues.

**Comment by AndrewCapon** [comment]
Hi @sastel,

I know this is a bit unrelated to the eMMC issue but did you get to the bottom of the DDR training not getting past 80%

Thanks

Andy

**Comment by sastel** [comment]
@AndrewCapon @sarah18-z I abandoned this effort without solving it. I ran into too many problems when trying to use Renode for embedded issues. Instead I'm debugging on hardware.

**Comment by AndrewCapon** [comment]
Hi @sastel,

Thanks very much for the info, I'm seeing the same issue in a different setup. Very annoying!

All the best

Andy

**Comment by sastel** [comment]
@AndrewCapon no problem. Not sure what your requirements are, but you could try launching your application .elf/.hex directly in Renode, bypassing the HSS. You won't be able to test the bootloader that way, but at least you can run your application. I used this to debug C++ exception handling on PolarFire. You may still encounter issues though with low level peripheral accesses.

**Comment by AndrewCapon** [comment]
@sastel 

Thanks, I will give it a go.

I have tried bypassing the DDR training, but that leads to further problems down the line. Here I think it is caused by some AXI4 streaming logic, well anyway if I remove that the problem goes away :)

I will dig deeper...

Many thanks

Andy

**Comment by sarah18-z** [comment]
Hello @AndrewCapon @sastel ,

I have some updates that might help you. I managed to get past the DDR insertion error by using HSS version 0.99.33, and it works correctly with Renode version 1.15.0. For DDR training, I simply disabled it in my HSS configuration.

Now, another problem has arisen: the HSS is unable to boot from the eMMC, and I am still working on it.
